### PR TITLE
Viewer fix: invalid html_widget resource paths

### LINKED ISSUE #665: Leaflet tiles not showing in web viewer
**Describe the bug**

Leaflet tiles are not showing in the web viewer.

**To Reproduce**

1. Install and load **leaflet**
2. Run `leaflet %>% addTiles()`
3. The plot opens up in the web viewer, there are zoom controls, but no map. It is all grey.

## Do you want to fix by self? (We hope your help!)

Yes

## (If yes,) what kind of help do you want? (e.g. Which file should I fix, Survey (related documents)

I'd love to help, but I don't know where to get started.

(If related)setting.json

```json
// R path for Linux
"r.rterm.linux": "/home/<user>/.local/bin/radian",

// R command line options (i.e: --vanilla)
"r.rterm.option": [],

// An optional encoding to pass to R when executing the file, i.e. 'source(FILE, encoding=ENCODING)'
"r.source.encoding": "UTF-8",

// Keeping focus when running
"r.source.focus": "editor",

// Use active terminal for all commands, rather than creating a new R terminal
"r.alwaysUseActiveTerminal": false,

// Use bracketed paste mode
"r.bracketedPaste": true,

// Enable R session watcher (experimental)
"r.sessionWatcher": true,

// Delay in milliseconds before sending each line to rterm (only applies if r.bracketedPaste is false)
"r.rtermSendDelay": 8,
```

**Expected behavior**

**Screenshots**

Here is what is looks like in VS Code:

image

Here is what is looks like in RStudio:

image

**Environment (please complete the following information):**

- OS: Linux (Fedora 34)
- VSCode Version: 1.56.2
- R Version: 4.0.5
- vscode-R version: 1.6.8

**Additional context**

When running a shiny application, the map renders just fine.

### PR DESCRIPTION
**Pull request by ElianHugh**
Resolves #665 

Adds a mutation observer to the webview body, which watches for changes in the DOM. Specifically, the filescheme `vscode-webview://` is tested for 'URLness' and (if true) is converted to an `https://` request.

This PR also defines some (minimal) [content security policies]() for the webview.

## Screenshot

image

## How can I check this pull request?
Prior to the PR, the following led to a blank viewer page, which should now be fixed:

```r
library(leaflet)
leaflet() |> addTiles()
```

Dynamic loading of resources is supported too:

```r
library(mapview)
mapview(breweries)
```

### CODE REVIEW

**Comment by renkun-ken** [comment]
Thanks for working on this! Works nicely already.

When I test the mapview example, it actually works, but I found the following error in the log:

```
fgb.js:103 Refused to connect to 'vscode-webview://8fe576af-ef9f-4237-9417-cd8966ff7eef/lib/breweries-0.0.1/breweries_layer.fgb' because it violates the following Content Security Policy directive: "default-src https: data: filesystem:". Note that 'connect-src' was not explicitly set, so 'default-src' is used as a fallback.

LeafletWidget.methods.addFlatGeoBuf @ fgb.js:103
fgb.js:103 Refused to connect to 'vscode-webview://8fe576af-ef9f-4237-9417-cd8966ff7eef/lib/breweries-0.0.1/breweries_layer.fgb' because it violates the document's Content Security Policy.
LeafletWidget.methods.addFlatGeoBuf @ fgb.js:103
fgb.js:103 Uncaught (in promise) TypeError: Failed to fetch
    at e.LeafletWidget.methods.addFlatGeoBuf (fgb.js:103)
    at Object.doRenderValue (leaflet.js:812)
    at Object.renderValue (leaflet.js:714)
    at Object.renderValue (htmlwidgets.js:886)
    at htmlwidgets.js:653
    at Array.forEach (<anonymous>)
    at forEach (htmlwidgets.js:55)
    at htmlwidgets.js:576
    at Array.forEach (<anonymous>)
    at forEach (htmlwidgets.js:55)
```

Not sure if it has any impact on the functionality. Anything we should do about this or we could ignore it?

File: src/session.ts [hunk 1]
```diff
@@ -269,10 +269,12 @@ export function openExternalBrowser(): void {
 
 export async function showWebView(file: string, title: string, viewer: string | boolean): Promise<void> {
     console.info(`[showWebView] file: ${file}, viewer: ${viewer.toString()}`);
+    console.log(file);
```
Comment (renkun-ken) [inline src/session.ts:272]: Is this redundant since the line above just prints `file` to console?

Comment (ElianHugh) [reply src/session.ts:272]: :facepalm: thanks for catching that

**Comment by ElianHugh** [comment]
> Thanks for working on this! Works nicely already.
> 
> When I test the mapview example, it actually works, but I found the following error in the log:
> 
> ```
> fgb.js:103 Refused to connect to 'vscode-webview://8fe576af-ef9f-4237-9417-cd8966ff7eef/lib/breweries-0.0.1/breweries_layer.fgb' because it violates the following Content Security Policy directive: "default-src https: data: filesystem:". Note that 'connect-src' was not explicitly set, so 'default-src' is used as a fallback.
> 
> LeafletWidget.methods.addFlatGeoBuf @ fgb.js:103
> fgb.js:103 Refused to connect to 'vscode-webview://8fe576af-ef9f-4237-9417-cd8966ff7eef/lib/breweries-0.0.1/breweries_layer.fgb' because it violates the document's Content Security Policy.
> LeafletWidget.methods.addFlatGeoBuf @ fgb.js:103
> fgb.js:103 Uncaught (in promise) TypeError: Failed to fetch
>     at e.LeafletWidget.methods.addFlatGeoBuf (fgb.js:103)
>     at Object.doRenderValue (leaflet.js:812)
>     at Object.renderValue (leaflet.js:714)
>     at Object.renderValue (htmlwidgets.js:886)
>     at htmlwidgets.js:653
>     at Array.forEach (<anonymous>)
>     at forEach (htmlwidgets.js:55)
>     at htmlwidgets.js:576
>     at Array.forEach (<anonymous>)
>     at forEach (htmlwidgets.js:55)
> ```
> 
> Not sure if it has any impact on the functionality. Anything we should do about this or we could ignore it?

Very weird, this doesn't occur on WSL, but does on Linux. Removing the CSP policies shows a fetch attempt to a non-existent filepath for some reason... I think it does effect the blue dots on the map, but is fixed if you type `mapviewOptions(fgb = FALSE)`. In looking into this I came across [this thread](https://github.com/r-spatial/mapview/issues/313) which appears to suggest this isn't just a VS Code thing.

**Comment by renkun-ken** [comment]
> Very weird, this doesn't occur on WSL, but does on Linux. Removing the CSP policies shows a fetch attempt to a non-existent filepath for some reason... I think it does effect the blue dots on the map, but is fixed if you type mapviewOptions(fgb = FALSE). In looking into this I came across this thread which appears to suggest this isn't just a VS Code thing.

Do you think if there is anything we could do about it?

**Comment by ElianHugh** [comment]
> Do you think if there is anything we could do about it?

image

I found this, so I guess it might be something that can be fixed. I'll have a look

**Review by renkun-ken — APPROVED** [review]
Works nicely now. 

LGTM

**Pull request merged**